File: 02-UML-Class-Diagrams/src/go/member.go
package main

import "fmt"

// Aggregation: Library has Members (they exist independently).
// Dependency: Member uses Book in Borrow() but doesn't store it.

type Member struct {
	MemberID string
	Name     string
}

func NewMember(memberID, name string) *Member {
	return &Member{MemberID: memberID, Name: name}
}

// Dependency: uses Book as parameter only
func (m *Member) Borrow(b *Book) {
	fmt.Printf("%s borrowed: %s\n", m.Name, b.Title)
}

File: 02-UML-Class-Diagrams/src/rust/member.rs
use crate::book::Book;

/// Dependency: Member uses Book in borrow() but doesn't own it
pub struct Member {
    member_id: String,
    name: String,
}

impl Member {
    pub fn new(member_id: &str, name: &str) -> Self {
        Member {
            member_id: member_id.to_string(),
            name: name.to_string(),
        }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn member_id(&self) -> &str {
        &self.member_id
    }

    // Dependency: borrows a reference to Book (doesn't store it)
    pub fn borrow_book(&self, book: &Book) {
        println!("{} borrowed: {}", self.name, book.title());
    }
}
